### ISSUE TITLE
Significant discrepancy between FedTHE ImageNet experiment results and paper-reported values

### ISSUE THREAD

**Issue by wenxinyang1026** [LINs-lab/FedTHE#6]
Hello，thanks for your excellent work! 
When running main.pywith FedTHE on ImageNet using the following configuration (obtained from run_exps.py):

--remote_exec False  --data imagenet  --data_dir datasets  --val_data_ratio 0  --train_data_ratio 1  --use_lmdb_data False  --partition_data_conf distribution=non_iid_dirichlet,non_iid_alpha=1.0,size_conf=1:1  --pin_memory True  --num_workers 0  --test_partition_ratio 0  --local_train_ratio 0.6  --eval_dataset test_loader  --test_ooc_perf_on_others True  --natural_shifted_imagenet_type imagenet_v2_matched-frequency  --weighted_sampling_mixed_test True  --arch resnet20  --group_norm_num_groups 2  --complex_arch master=resnet20,worker=resnet20  --w_conv_bias False  --w_fc_bias False  --freeze_bn False  --freeze_bn_affine False  --resnet_scaling 1  --is_personalized True  --is_corrupted True  --corr_severity 5  --corr_seed 5  --personalization_scheme method=THE  --n_personalized_epochs 1  --personal_lr 0.01  --with_BN False  --fedrep_personal_layers 1  --regularized_factor 1.0  --is_dynamic_lambda False  --fedrod_personal_layers 1  --t3a_filter_k 50  --rep_len 256  --is_rep_history_reused False  --THE_steps 20  --THE_alpha 0.1  --THE_beta 0.1  --drfa_sync_gap 1  --drfa_lambda_lr 0.01  --comm_buffer_size 300  --n_comm_rounds 100  --early_stopping_rounds 0  --local_n_epochs 5  --reshuffle_per_epoch False  --batch_size 128  --n_clients 20  --participation_ratio 1.0  --n_master_sampled_clients 20  --fl_aggregate scheme=federated_average  --train_fast False  --lr 0.01  --global_lr 1  --lr_scheduler MultiStepLR  --lr_decay 0.1  --lr_patience 10  --lr_scaleup False  --lr_warmup False  --lr_warmup_epochs 5  --lr_warmup_epochs_upper_bound 150  --adam_beta_1 0.9  --adam_beta_2 0.999  --adam_eps 1e-08  --optimizer sgd  --momentum_factor 0.9  --use_nesterov False  --weight_decay 0.0005  --drop_rate 0.5  --weighted_beta 0  --weighted_gamma 0  --densenet_growth_rate 12  --densenet_bc_mode False  --densenet_compression 0.5  --wideresnet_widen_factor 4  --mlp_num_layers 2  --mlp_hidden_size 128  --rnn_n_hidden 200  --rnn_n_layers 2  --rnn_bptt_len 35  --rnn_clip 0.25  --rnn_use_pretrained_emb True  --rnn_tie_weights True  --rnn_weight_norm False  --transformer_n_layers 6  --transformer_n_head 8  --transformer_dim_model 512  --transformer_dim_inner_hidden 2048  --transformer_n_warmup_steps 4000  --same_seed_process False  --manual_seed 7  --evaluate False  --display_log False  --summary_freq 256  --timestamp 1759042687  --track_time True  --track_detailed_time False  --display_tracked_time True  --checkpoint ./data/checkpoint  --save_all_models False  --python_path /opt/conda/bin/python  --world 0,0,0,0,0,0,1,1,1,1,1,2,2,2,2,2,3,3,3,3,3  --on_cuda True  --hostfile env/hostfile  --mpi_path /.openmpi  --experiment debug  --job_name default  --job_id /tmp/jobrun_logs_1759042355  --script_path exps/exp_imagenet_resnet.py  --num_jobs_per_node 1  --wait_in_seconds_per_job 30

The experimental results at round 100 are:
{
  "measurement": "runtime",
  "time": "2025-09-29 03:03:53",
  "method": "THE",
  "round": 100,
  "global_top1": 61.63194444444444,
  "top1": 67.11727905273438,#test
  "corr_top1": 6.716606140136719,  # imagenet_a
  "ooc_top1": 16.38376808166504,   # imagenet_r
  "natural_shift_top1": 3.067927122116089,  # imagenet_v2
  "ooc_corr_top1": 0.0,
  "mixed_top1": 42.56490707397461,#mixture of test, corr_test, imagenet-a, imagenet-r, and imagenet-v2
  "split": "test"
}

However, these results differ significantly from the values reported in the paper:
Paper-reported top1: 86.56 ±0.64
imagenet_a: 6.76 ±0.40
imagenet_r: 59.62 ±4.43
imagenet_v2: 25.26 ±0.19
mix: 41.41
Could you please help identify the potential reasons for these large discrepancies? Are there any critical configuration differences or implementation details that might explain this variance?

**Comment by LiangzeJiang** [comment]
Hello,

Thank you for your interest in our work.

Could you try --n_comm_rounds 300? As mentioned in Appendix C.1, the experiments on ImageNet are done with 300 communication rounds.

Let us know if the discrepancy still persists.

**Comment by wenxinyang1026** [comment]
Thank you for your previous suggestion. I increased the n_comm_roundsto 300, and the results are now much closer to those reported in the paper. Here are the new results at round 300:
{
  "method": "THE",
  "round": 300,
  "ooc_top1": 20.643869400024414,       # Actually corresponds to ImageNet-R
  "natural_shift_top1": 61.049232482910156, # Actually corresponds to ImageNet-V2
}
While most results now align with the paper, there is a clear inconsistency regarding the Out-of-Distribution (OOD) datasets ​​ImageNet-V2​​ and ​​ImageNet-R​​.
In my results:
ooc_top1(20.64) corresponds to ​​ImageNet-R​​
natural_shift_top1(61.05) corresponds to ​​ImageNet-V2​​
However, the paper reports the opposite:
​​ImageNet-R​​: 59.62 ±4.43
​​ImageNet-V2​​: 25.26 ±0.19

I have traced the variable assignment in the code to understand this swap:
In base_worker.py, within the _init_train_process()function, the local_data_loadersare created.
The metric ooc_top1 is linked to local_data_loaders["natural_shift_test_r"].
The metric natural_shift_top1 is linked to local_data_loaders["natural_shift_test"].
Tracing back to create_dataset.py in the intra_client_data_partition_and_create_dataloaders() function reveals the following mapping:
local_data_loaders["natural_shift_test_r"] is assigned to data_loader_local_r_te(which is the ​​ImageNet-R​​ dataset).
local_data_loaders["natural_shift_test"] is assigned to data_loader_local_v2_te(which is the ​​ImageNet-V2​​ dataset).
This confirms that the variable names for these two datasets are swapped relative to the paper's reporting convention.

Could you please clarify the reason for this discrepancy? Thank you for your help.

**Comment by LiangzeJiang** [comment]
Thanks for identifying this! I think there might indeed be a typo in the paper swapping the results of the two datasets, because ImageNet-R should be inherently more challenging than ImageNet-V2 (so the accuracy on R should be lower than V2).
I will double-check when I get to it and correct the typo in the paper if necessary.
